service/src/main/java/bio/terra/workspace/service/resource/ValidationUtils.java

```diff
@@ -65,6 +65,9 @@
   public static final Pattern RESOURCE_NAME_VALIDATION_PATTERN =
       Pattern.compile("^[a-zA-Z0-9][-_a-zA-Z0-9]{0,1023}$");
 
+  public static final Pattern GIT_CLONE_URL_PATTERN =
```

Comment:
This pattern of allowing users to pass arbitrary URLs and having WSM call them (which we would need to do for access validation) scares me a little from a security perspective.

What we did for TDR snapshots was to configure a list of allowed instances (e.g. for dev WSM, allow dev TDR with a known URL) and have requests indicate which instance they were referring to. Could we do something similar here?

E.g., we would explicitly only allow git references from [Github, Gitlab, Azure, Github enterprise, GCP cloud source repository, others added as needed]. The "createGitReference" request would have fields like "instance" with enum values like "GITHUB", "GITLAB", ..., plus fields like "repoName", and WSM could construct the URL from the combination of the two. That way, users couldn't force WSM to make calls to arbitrary URLs.

Reply:
I don't think we could enumerate and then keep up with all the git origins out there. For example, what about internal company servers that we wouldn't know about? I like the idea of passing URL parts separately. Plus it expands to tags/releases and commits naturally.

If we did have a fixed list, we'd want an API to return it so the UI could have a dropdown for the entries.

Reply:
Is WSM directly making requests on the URL? We are aiming for an enterprise github, so github.com might not be one the allowed list.

A whitelist probably makes sense in the fullness of time. I wonder if there is a wider issue of controlling external access from a workspace that is more in the realm of policy and security perimeters rather than point solutions for single URLs.

Reply:
I was going to just have a github referenced resource but then changed to a generic git url because i was trying to handle as many cases as possible (e.g. servers that we don't know about) but security concern is a good point. 

Maybe we can start with constructing the url for github enterprise case first and add to the api as needed then? 

an enterprise github url would still look like https://github.com/organization/project.git or git@github.com:organization/project.git so I would need user input for organization and project name and whether they want ssh url or https url.

Reply:
Agreed, the number + variation of sources makes it messier to maintain a list like this than it is in the TDR case. Whitelisting URLs also sgtm if we'd rather the API still take in whole URLs instead of instance identifiers.

IIUC yes, the current design would have WSM directly making requests to the provided URL using the user's credentials to validate a user's repo access before creating the reference (though we don't right now since validation isn't implemented). I think there's probably a broader policy question of limiting the things a workspace can reach, but I think this is a much more local security issue of "don't let users make your service call arbitrary URLs".

Reply:
right now the user case I'm imaging is that the jupyter notebook uses the git repo references to clone some workspaces in the notebook instance. I'm not sure if there's value to directly make GIT operation in WSM.

Reply:
Thanks for all the suggestions. I listed the three options here https://docs.google.com/document/d/13pEGiu9OOk_zhdpi-LPJhwjPEbUVdeQQT43PqLhc2uE/edit?resourcekey=0-qrxEly_ybyUr4d9t68fT6w#bookmark=id.xuuyb98gkeza. Would we like to settle on option 3: having a set of allowlisted hostname?

Reply:
I think doing the allow list is best, even though it means more work to maintain it.
Is the hostname sufficient for the allow list? I think so, but I'm not a security expert.

Implementing allow list will be a separate piece of work and need some design around how we maintain and propagate that list. Is it deployment wide or a policy that propagates down folders? Or something else? Maybe need some product requirements.

Reply:
added an allowlist of hostname as an external/gitreporeferenceconfiguration. not sure if that's the right directory to put it. please take a look.

so for https:// and ssh:// url, we parse it to a uri and then check hostname. If it's a git@ ssh uri, we use regex to get the host name and then check whether it is in the allowlist.